### PR TITLE
Fix list equal for empty offset list array

### LINKED ISSUE #1817: List equality method should work on empty offset ListArray
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`list_equal` currently fails on ListArrays with empty offsets. This causes interop test failure in https://github.com/apache/arrow/pull/13219:

```
FAILED TEST: nested_large_offsets Java producing,  Rust consuming
```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1817.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/array/equal/list.rs [hunk 1]
```diff
@@ -73,11 +73,19 @@ pub(super) fn list_equal<T: OffsetSizeTrait>(
     // however, one is more likely to slice into a list array and get a region that has 0
     // child values.
     // The test that triggered this behaviour had [4, 4] as a slice of 1 value slot.
-    let lhs_child_length = lhs_offsets[lhs_start + len].to_usize().unwrap()
-        - lhs_offsets[lhs_start].to_usize().unwrap();
+    let lhs_child_length = if len == 0 {
```
Comment (HaoYang670) [inline arrow/src/array/equal/list.rs:76]: What about just add 
```rust
// Zero length list arrays are always equal
if len == 0 {
    return true;
}
```
at the beginning of the function?

File: arrow/src/array/equal/list.rs [hunk 2]
```diff
@@ -73,11 +73,19 @@ pub(super) fn list_equal<T: OffsetSizeTrait>(
     // however, one is more likely to slice into a list array and get a region that has 0
     // child values.
     // The test that triggered this behaviour had [4, 4] as a slice of 1 value slot.
```
Comment (HaoYang670) [inline arrow/src/array/equal/list.rs:75]: nit: We should also update the docs. Because we add an edge case for zero-length array

File: arrow/src/array/equal/mod.rs [hunk 3]
```diff
@@ -629,6 +629,39 @@ mod tests {
         test_equal(&a, &b, false);
     }
 
+    #[test]
```
Comment (HaoYang670) [inline arrow/src/array/equal/mod.rs:632]: Should we also test that 2 zero-length arrays with different (offset, null-buffer, childdata) are equal?

**Comment by HaoYang670** [comment]
I guess this PR is related to #1620.

**Review by tustvold — APPROVED** [review]
